lib/common/src/DevMode.php

```diff
@@ -62,9 +62,15 @@ public static function hasExemptionForNode(DOMNode $node)
      */
     public static function isExemptFromValidation(DOMNode $node)
     {
-        $document = $node instanceof Document ? $node : $node->ownerDocument;
+        $document = $node->ownerDocument;
+        if (! $node->ownerDocument instanceof Document) {
+            $document = Document::fromNode($node);
+        }
 
-        if (! $document) {
+        /**
+         * @var Document $document
+         */
+        if ($node === $document->documentElement && ! $document->hasInitialAmpDevMode()) {
```

Comment:
There are a few places in the code that forgo `DevMode::isExemptFromValidation()` and use `DevMode:: hasExemptionForNode()` instead (to duplicate calls to `DevMode::isActiveForDocument()`).

You should check these and make sure they behave correctly still, or adapt them as needed.

Reply:
The only change here is to determine whether the `html` element should be validated or not; the behavior of the class is the same otherwise.

I haven't seen any adverse effects of determining the dev mode for nodes with this change.

Reply:
I rescind my comment above; please see changes in https://github.com/ampproject/amp-wp/pull/4440/commits/3081a0a5de6e45c3d01ea0c9059f9914dc37128b.